homeassistant/components/hlk_sw16.py

```diff
@@ -0,0 +1,294 @@
+    # Allow platform to specify function to register new unknown devices
+    from hlk_sw16 import create_hlk_sw16_connection
+    hass.data[DATA_DEVICE_REGISTER] = {}
+    switches = config[DOMAIN][CONF_SWITCHES]
+
+    @callback
+    def event_callback(event):
+        """Handle incoming HLK-SW16 events.
+
+        HLK-SW16 events arrive as dictionaries of varying content
+        depending on their type. Identify the events and distribute
+        accordingly.
+        """
+        _LOGGER.debug('event: %s', event)
+
+    host = config[DOMAIN][CONF_HOST]
+    port = config[DOMAIN][CONF_PORT]
+
+    @callback
+    def reconnect(exc=None):
+        """Schedule reconnect after connection has been unexpectedly lost."""
+        # Reset protocol binding before starting reconnect
+        SW16Command.set_hlk_sw16_protocol(None)
+
+        async_dispatcher_send(hass, SIGNAL_AVAILABILITY, False)
+
+        # If HA is not stopping, initiate new connection
+        if hass.state != CoreState.stopping:
+            _LOGGER.warning('disconnected from HLK-SW16, reconnecting')
+            hass.async_create_task(connect())
+
+    async def connect():
+        """Set up connection and hook it into HA for reconnect/shutdown."""
+        _LOGGER.info('Initiating HLK-SW16 connection')
+
+        try:
+            with async_timeout.timeout(CONNECTION_TIMEOUT,
+                                       loop=hass.loop):
+                transport, protocol = await create_hlk_sw16_connection(
+                    host=host,
+                    port=port,
+                    event_callback=event_callback,
+                    disconnect_callback=reconnect,
+                    loop=hass.loop,
+                    logger=_LOGGER)
+                _LOGGER.info(transport)
+
+        except (ConnectionRefusedError, TimeoutError, OSError,
+                asyncio.TimeoutError) as exc:
+            reconnect_interval = config[DOMAIN][CONF_RECONNECT_INTERVAL]
+            _LOGGER.exception(
+                "Error connecting to HLK-SW16, reconnecting in %s",
+                reconnect_interval)
+            # Connection to HLK-SW16 device is lost, make entities unavailable
+            async_dispatcher_send(hass, SIGNAL_AVAILABILITY, False)
+
+            hass.loop.call_later(reconnect_interval, reconnect, exc)
+            return
+
+        comps = {'switch': []}
+        for switch in switches.items():
+            comps['switch'].append(switch)
+
+        # Load platforms
+        for comp_name, comp_conf in comps.items():
+            if comp_conf:
+                load_platform(hass, comp_name, DOMAIN, {DOMAIN: comp_conf},
+                              config)
+
+        # Bind protocol to command class to allow entities to send commands
+        SW16Command.set_hlk_sw16_protocol(protocol)
+
+        # There is a valid connection to a HLK-SW16 device now so
+        # mark entities as available
+        async_dispatcher_send(hass, SIGNAL_AVAILABILITY, True)
```

Comment:
This is more an FYI. There won't be an error due to this, as far as I can see.

Reply:
I've removed this, what should I do for the other areas where `SIGNAL_AVAILABILITY` is used? Should that be device specific or entity/relay specific?

Reply:
One signal can be attached to multiple receivers, eg multiple entities. If there's granularity in the availability eg if only one of the relays could go offline, than it's good to keep that level of granularity.

Reply:
With how the device is designed it's not possible for relays to go offline individually(it's essentially 1 MCU that handles the control protocol for the 16 attached physical relays). So if the device goes offline all of the 16 relay switches it has should also go offline at the same time. What should I do here since that granularity doesn't seem to be needed?

Reply:
We still want to send the signal when the device goes offline. So connect all entities to the same signal for availability.